internals-js/src/specs/__tests__/sourceSpec.test.ts

```diff
@@ -1,7 +1,309 @@
+  parseURLPathTemplate,
+  getURLPathTemplateVars,
+} from '../index';
 
 describe('SourceSpecDefinition', () => {
   it('should export expected identity URL', () => {
     expect(sourceIdentity).toBe('https://specs.apollo.dev/source');
   });
 });
+
+function parseSelectionExpectingNoErrors(selection: string) {
+  const ast = parseJSONSelection(selection)!;
+  expect(ast.errors).toEqual([]);
+  return ast;
+}
+
+describe('parseJSONSelection', () => {
+  it('parses simple selections', () => {
+    expect(parseSelectionExpectingNoErrors('a')!.type).toBe('Selection');
+    expect(parseSelectionExpectingNoErrors('a b')!.type).toBe('Selection');
+    expect(parseSelectionExpectingNoErrors('a b { c }')!.type).toBe('Selection');
+    expect(parseSelectionExpectingNoErrors('.a')!.type).toBe('Selection');
+    expect(parseSelectionExpectingNoErrors('.a.b.c')!.type).toBe('Selection');
+  });
+
+  const complexSelection = `
+    # Basic field selection.
+    foo
+
+    # Similar to a GraphQL alias with a subselection.
+    barAlias: bar { x y z }
+
+    # Similar to a GraphQL alias without a subselection, but allowing for JSON
+    # properties that are not valid GraphQL Name identifiers.
+    quotedAlias: "string literal" { nested stuff }
+
+    # Apply a subselection to the result of extracting .foo.bar, and alias it.
+    pathAlias: .foo.bar { a b c }
+
+    # Nest various fields under a new key (group).
+    group: { foo baz: bar { x y z } }
+
+    # Get the first event from events and apply a selection and an alias to it.
+    firstEvent: .events.0 { id description }
+
+    # Apply the { nested stuff } selection to any remaining properties and alias
+    # the result as starAlias. Note that any * selection must appear last in the
+    # sequence of named selections, and will be typed as JSON regardless of what
+    # is subselected, because the field names are unknown.
+    starAlias: * { nested stuff }
+  `;
+  // TODO Improve error message when other named selections accidentally follow
+  // a * selection.
+
+  it('parses a multiline selection with comments', () => {
+    expect(parseSelectionExpectingNoErrors(complexSelection)!.type).toBe('Selection');
+  });
+
+  describe('getSelectionOutputShape', () => {
+    it('returns the correct output shape for a simple selection', () => {
+      const ast = parseSelectionExpectingNoErrors('a');
+      if (!ast) {
+        throw new Error('Generic parse failure for a');
+      }
+      expect(getSelectionOutputShape(ast)).toEqual({
+        a: 'JSON',
+      });
+    });
+
+    it('returns the correct output shape for a complex selection', () => {
+      const ast = parseSelectionExpectingNoErrors(complexSelection);
+      if (!ast) {
+        throw new Error('Generic parse failure for ' + complexSelection);
+      }
+      expect(getSelectionOutputShape(ast)).toEqual({
+        foo: 'JSON',
+        barAlias: {
+          x: 'JSON',
+          y: 'JSON',
+          z: 'JSON',
+        },
```

Comment:
The output shape of a `JSONSelection` string is a tree of nested JSON dictionaries mirroring the structure of a GraphQL query, including scalar properties as leaf nodes. The ultimate goal is to validate this structure against object fields to ensure all fields in the schema are handled by some `JSONSelection` mapping.

These `'JSON'` leaf values could have been any placeholder value, but I wanted to emphasize we can't assume much about the types of leaf fields within `JSONSelection` output shapes, except that they must be some JSON value (including possibly `null` for nullable fields).